enhancements/installer/connected-onprem-assisted-installer.md

```diff
@@ -0,0 +1,248 @@
+---
+title: assisted-installer-on-prem
+authors:
+  - "@mhrivnak"
+reviewers:
+  - "@avishayt"
+  - "@hardys"
+  - "@dhellmann"
+  - "@beekhof"
+approvers:
+  - "@crawford"
+  - "@abhinavdahiya"
+  - "@eparis"
+creation-date: 2020-07-23
+last-updated: 2020-07-23
+status: implementable
+see-also:
+  - "/enhancements/installer/connected-assisted-installer.md"  
+---
+
+# Assisted Installer On-Prem
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The assisted installer, [previously proposed and
+accepted](connected-assisted-installer.md), includes a service running on
+cloud.redhat.com to which agents connect and wait for orders on how to
+A) provision the host they are running on, and B) join or form a cluster.
+
+This enhancement proposes to enable a user to run that "assistant" service
+on-premise, inside their own network and on their own infrastructure, for
+one-time-use to create a new cluster.
+
+## Motivation
+
+Many users will not want a service that can provision hosts inside their
+network to be run by a third party outside of their network. Reasons can
+include trust, control, and reproducibility. Those users will prefer to run the
+service on-premise.
+
+### Goals
+
+* Users can run the entire assisted install workflow with software that runs inside their network.
+
+### Non-Goals
+
+* Disconnected and air-gapped environments are not included in this enhancement.
+* Adding Nodes on day 2 will be addressed separately.
+* Utilizing a local container registry for mirrored content will be addressed in a future enhancement.
+
+## Proposal
+
+A user will visit cloud.redhat.com to download a live ISO with which they can
+run the assisted installer on-premise. The RHCOS-based ISO will include:
+
+* the user's pull secret
```

Comment:
Is this the only customer-specific information in the ISO? Can we remove this and just ask for the information at runtime? That would allow us to distribute the same ISO to everyone versus having to run a service which can perform these modifications.

Reply:
That depends on how easy it would be to provide it at runtime. The idea is that the ISO starts one or more containers to run the assisted installer, but just obtaining those container images requires the pull secret. Would we be better-off baking a UI into the ISO somehow that helps the user obtain their pull secret? Or can we bake the container images themselves into the ISO? How would we accomplish either?

Reply:
but that means interacting with the iso during boot, which isn't neccesarily trivial, and the interfaces for that don't necessarily support copy/paste well?

Reply:
This ISO is for the local copy of the service, right? It's not the ISO that runs on the hosts used to form the cluster? So to interact with it, the user could use a web browser or remote login, right?

Why wouldn't the container images we need to run the service be embedded in the ISO itself?

Reply:
There is not currently a way to embed container images into an RHCOS ISO. We can only embed ignition.

I agree that requiring input at boot time would not be a good UX and is worth avoiding.

If we want to enable the user to run a live ISO that includes the assisted installer, we can show them how to perform the following steps. Perhaps in the future the ISO could be generated automatically on-demand from cloud.redhat.com.

1. Download an RHCOS live ISO image.
2. Download a pre-baked ignition file that runs the assisted-installer using podman.
3. Add your pull secret to the ignition file.
4. Run `coreos-installer iso embed ...`
5. Boot the resulting ISO.

Steps 2 and 3 might be simple to combine; cloud.redhat.com could automatically add the current user's pull secret to the ignition blob before they download it.

Reply:
No pull secret is required to download the openshift-install binary in the current, standard OpenShift installation flow.

This service is equivalent - it's a thing that asks you for your pull secret when you execute it. (Or offers the option for you to log in to your Red Hat account and fetches your pull secret automatically)

Why can't we apply the same thinking - make the content for the assisted installer available without a pull secret?

Reply:
I like that thinking. The difference that matters most is that the openshift-install command is just a binary, and it does not include other content that normally requires a pull secret. Running the assisted-installer on-prem however requires the user to retrieve:
* RHCOS
* multiple container images built on a RHEL base image
* a supported postgresql container image

If there's a legal way for us to enable users to download those artifacts without having agreed to RH terms of service (that's the main purpose of the pull secret), then we could simplify the workflow as you suggest. I'm just not aware of that being a possibility, and it seems unlikely based on my zero-legal-expertise understanding of why we require pull secrets.